File: fsharp/binary-search-tree/BinarySearchTree.fs
module BinarySearchTree

type Node = { left: Node option; value: int; right: Node option }

let left node  = node.left
let right node = node.right
let data node = node.value

let private singleton value = { left = None; right = None; value = value }

let rec private insert newValue (tree: Node) =
    let loop newValue parentOpt =
        match parentOpt with
        | Some parent -> Some (insert newValue parent)
        | None -> Some (singleton newValue)

    if newValue <= tree.value then
        { tree with left = loop newValue tree.left }
    else
        { tree with right = loop newValue tree.right }

let sortedData tree = 
    let rec loop = 
        function
        | Some node -> loop node.left @ [node.value] @ loop node.right
        | None -> []

    loop (Some tree)

let create = 
    function
    | []    -> failwith "Cannot create tree from empty list."
    | x::xs -> List.fold (fun acc elem -> insert elem acc) (singleton x) xs

File: elixir/binary-search-tree/lib/binary_search_tree.ex
defmodule BinarySearchTree do
  def new(data), do: %{data: data, left: nil, right: nil}

  def insert(nil, data), do: new(data)

  def insert(tree, data) do
    cond do
      data <= tree.data -> %{tree | left: insert(tree.left, data)}
      true -> %{tree | right: insert(tree.right, data)}
    end
  end

  def in_order(nil), do: []

  def in_order(tree) do
    in_order(tree.left) ++ [tree.data] ++ in_order(tree.right)
  end
end
